### PR TITLE
runtime: support bpf_ktime_get_boot_ns

### LINKED ISSUE #608: [BUG]  llvm-jit cannot find bpf helper when bpftime trying to attach to running process
**Describe the bug**
bpftime failed to attach bpf programs to a running program from agent side.
the bpf pgorams comes from "bpftime load " subcommand from server side.

**To Reproduce**

./my_app
sudo bpftime load bpftrace simple.bt
sudo bpftime attach  $(pgrep my_app)
sudo bpftime detach

the content of simple.bt

#include "/lib/my_app.h"

uprobe:/lib/my_app.so:my_func
{
        $msg = (struct my_message_s *)arg1;
        $serv_id = $msg->hdr.serv;
        $meth_id = $msg->hdr.method;

        if ( $serv_id == 1 && $meth_id == 11)  {
                @start_at["my_func"] = nsecs;
                printf("%-6d\n", pid);
        }
}

**Expected behavior**
works the same as  bpftrace simple.bt

**Screenshots**
.bpftime/runtime.log
[2026-07-24 03:12:54][info][17624] Initialize syscall server
[2026-07-24 03:12:54][info][17624] Global shm constructed. shm_open_type 3 for bpftime_maps_shm
[2026-07-24 03:12:54][info][17624] Global shm initialized
[2026-07-24 03:12:54][info][17624] bpftime-syscall-server started
[2026-07-24 03:13:00][info][17628] Initialize syscall server
[2026-07-24 03:13:00][info][17628] Global shm constructed. shm_open_type 3 for bpftime_maps_shm
[2026-07-24 03:13:00][info][17628] Global shm initialized
[2026-07-24 03:13:00][info][17628] bpftime-syscall-server started
[2026-07-24 03:13:01][warning][17636] Calling original ioctl: 2 21523 136024452079592
[2026-07-24 03:13:01][info][17628] Created uprobe/uretprobe perf event handler, module name /lib/my_app.so, offset 2fc40
[2026-07-24 03:13:09][info][17644] bpf_attach_ctx constructed
[2026-07-24 03:13:09][info][17644] Register attach-impl defined helper bpf_get_func_arg, index 183
[2026-07-24 03:13:09][info][17644] Register attach-impl defined helper bpf_get_func_ret_id, index 184
[2026-07-24 03:13:09][info][17644] Register attach-impl defined helper bpf_get_retval, index 186
[2026-07-24 03:13:09][info][17644] Initializing agent..
[2026-07-24 03:13:09][info][17644] bpftime: shm epoch_seq = 28
[2026-07-24 03:13:09][error][17644] LLVM-JIT: failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to load program helpers for prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate bpf prog handler 10: -1
[2026-07-24 03:13:09][info][17644] Executable path: /root/my_app
[2026-07-24 03:13:09][error][17644] LLVM-JIT: failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to load program helpers for prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate bpf prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate prog handler 10 when instantiating link handler 14: -1
[2026-07-24 03:13:09][error][17644] LLVM-JIT: failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to load program helpers for prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate bpf prog handler 10: -1
[2026-07-24 03:13:09][error][17644] LLVM-JIT: failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to compile: Ext func not found: _bpf_helper_ext_0125
[2026-07-24 03:13:09][error][17644] Failed to load program helpers for prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate bpf prog handler 10: -1
[2026-07-24 03:13:09][error][17644] Unable to instantiate prog handler 10 when instantiating link handler 14: -1
[2026-07-24 03:13:09][info][17644] agent ipc: listening (abstract) for pid 17611
[2026-07-24 03:13:09][info][17644] Attach successfully
[2026-07-24 03:13:47][info][17648] Detaching..
[2026-07-24 03:13:47][info][17649] Detaching..

**Desktop (please complete the following information):**

OS: ubuntu 24.04.4 LTS (Noble Numbat)
gcc version 13.3.0 (Ubuntu 13.3.0-6ubuntu2~24.04.1)
Ubuntu clang version 18.1.3 (1ubuntu1)
bpftrace v0.20.2

**Additional context**
commit id: 
005635ba58160a35e7e9f986bdbf37daf466508a

build step:
cmake -Bbuild -DCMAKE_BUILD_TYPE=Release -DBPFTIME_ENABLE_LTO=NO -DBPFTIME_LLVM_JIT=YES
cmake --build build --config Release --target install

no errow with the following chagnes:
#include "/lib/my_app.h"

uprobe:/lib/my_app.so:my_message
{
        printf("%-6d\n", pid);
}

### PR DESCRIPTION
**Pull request by yunwei37**
## Summary

- implement and register helper ID 125 (`bpf_ktime_get_boot_ns`)
- use `CLOCK_BOOTTIME`, with `CLOCK_MONOTONIC` as a portability fallback
- exercise the helper through the LLVM JIT helper regression program

Fixes #608

## Validation

- `cmake -B build -DCMAKE_BUILD_TYPE=Release -DBPFTIME_ENABLE_LTO=NO -DBPFTIME_LLVM_JIT=YES -DBPFTIME_ENABLE_UNIT_TESTING=1`
- `cmake --build build --target bpftime_runtime_tests -j$(nproc)`
- `./build/runtime/unit-test/bpftime_runtime_tests 'Test helpers'` (1 test case, 5 assertions passed)
- full runtime test binary: helper regression passed; 75/77 test cases passed, with the two remaining failures requiring host filesystem/kernel BPF privileges (`Test shm progs attach`, `Test tail calling from userspace to kernel`)

### CODE REVIEW

**Comment by yunwei37** [comment]
Exact-head review is complete for `8c6e281`.

The first pass identified a runtime edge case where `CLOCK_BOOTTIME` could be available at compile time but rejected by the running kernel. Commit `8c6e281` resolves it by zero-initializing the result and falling back to `CLOCK_MONOTONIC` on failure. Two final independent reviews found no blocking issues: helper ID/signature/registration are correct, the LLVM JIT regression covers the original unresolved-helper path, and the two-file 20-line scope remains minimal.

The focused test was rebuilt and rerun on the final head: 1 test case and 5 assertions passed. CI has now reached a terminal state with 141 successful checks, 2 expected skips, and no failed checks. The two GPU jobs were cancelled after waiting for unavailable self-hosted runners without executing any repository steps. The PR is mergeable and has no unresolved review threads; the remaining decisions are GPU-runner availability and maintainer review.

AI-generated response; a maintainer will review and follow up later

**Comment by yunwei37** [comment]
Final pushed head `8c6e281b768e9c09b85c948fc1b6565d71f7f961` passed both required independent read-only review gates.

- Correctness: helper ID 125, runtime ABI signature, registration, `CLOCK_BOOTTIME` semantics, and `CLOCK_MONOTONIC` portability fallback are correct.
- Regression coverage: the existing helper BPF asset now emits and executes `call 0x7d`, covering the original unresolved-helper failure without a new fixture.
- Scope/minimality: 2 files, +20/-0 (production +18, test +2); no new API/config/docs/abstraction and no practical safe reduction remains.
- Blocking findings: none.

Copilot review threads/comments were also checked; none exist on this PR.

**Comment by yunwei37** [comment]
Final exact-head review gate completed for `45f8f66c7ee96798c3a89804916696741d10dcbb`.

- Dedicated review: PASS; local/origin/PR refs match, the worktree is clean, and the exact LLVM helper test passed (1 case, 5 assertions).
- Independent read-only review: PASS; helper ID 125, `CLOCK_BOOTTIME` semantics, `CLOCK_MONOTONIC` portability fallback, registration ABI, and fixture relocation/execution were verified.
- Final net diff remains only 2 files, `+20/-0`; merging current master removed the unrelated CUDA compilation failure without adding it to the PR diff.
- No blocker or further material safe reduction remains.

File: runtime/src/bpf_helper.cpp [hunk 1]
```diff
@@ -315,6 +315,18 @@ uint64_t bpftime_ktime_get_coarse_ns(uint64_t, uint64_t, uint64_t, uint64_t,
...
+uint64_t bpftime_ktime_get_boot_ns(uint64_t, uint64_t, uint64_t, uint64_t,
+				   uint64_t)
+{
+	timespec spec{};
+#ifdef CLOCK_BOOTTIME
+	if (clock_gettime(CLOCK_BOOTTIME, &spec) == 0)
+		return spec.tv_sec * (uint64_t)1000000000 + spec.tv_nsec;
+#endif
+	clock_gettime(CLOCK_MONOTONIC, &spec);
+	return spec.tv_sec * (uint64_t)1000000000 + spec.tv_nsec;
```
[the remark this answers is not in the document]
Comment (tgoodwin) [reply runtime/src/bpf_helper.cpp:327]: Kept intentionally. The `timespec{}` zero-initialization makes the fallback failure result deterministic, and the BPF helper ABI has no separate error channel: returning `0` is the explicit failure convention here. Adding a log would make a hot-path time helper noisy, while an explicit `if` would preserve exactly the same observable result with more code.

File: runtime/unit-test/assets/helpers.bpf.c [hunk 2]
```diff
@@ -17,5 +17,7 @@ SEC("uprobe")
 int print_and_add1(struct data *d, int sz) {
 	bpf_printk("print_and_add1\n");
 	bpf_printk("print_and_add1: %d\n", sz);
+	if (!bpf_ktime_get_boot_ns())
+		return 0;
```
[the remark this answers is not in the document]
Comment (tgoodwin) [reply runtime/unit-test/assets/helpers.bpf.c:21]: Kept as the smallest regression signal. This fixture's purpose is to prove helper ID 125 is registered, parsed, and executed by the existing LLVM regression path; the surrounding test requires the program result to be exactly `23`, so a zero result fails deterministically. Calling twice does not remove the zero ambiguity (`0 >= 0`), and extra printk plumbing would enlarge this focused test.

File: runtime/src/bpf_helper.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgoodwin) [reply runtime/src/bpf_helper.cpp:327]: Not extracted to keep this two-file helper addition minimal. The two expressions are adjacent branches of the same helper and use the identical conversion; a utility would add another symbol/abstraction without reducing the changed behavior or LOC.
